engineering_tools/carma_script_extensions/swarm

```diff
@@ -0,0 +1,352 @@
+        exit 1
+    fi
+
+    if [ ! -z "$(echo $TAG | grep :)" ]; then
+        local IMAGE_NAME="$TAG"
+    else
+        local IMAGE_NAME="$USERNAME/carma-config:$TAG"
+    fi
+
+    if [ "$(docker images | grep $USERNAME/carma-config | grep $TAG )" ]; then
+        echo "Found image, applying to carma-config-volume"
+        docker run --rm --name carma-config-data -v carma-config-data:/opt/carma/vehicle/config -d $USERNAME/carma-config:$TAG
+    else
+        echo "Could not find $USERNAME/carma-config:$TAG in local images"
+    fi
+}
+
+
+swarm-config__edit() {
+    echo "Opening carma-config-data volume shell with read/write privileges..."
+
+    if [ ! -z "$(docker volume ls | grep carma-config-data)" ]; then
+        echo "Found carma-config-data volume"
+    else
+        echo "Missing carma-config-data volume, add with \"carma swarm update-config <image tag>\""
+        exit 1
+    fi
+
+    local carma_base=$(__get_image_from_config carma-base:)
+    if [[ -z $carma_base ]]; then
+        __pull_newest_carma_base
+        carma_base=$(__get_most_recent_carma_base)
+    fi
+
+    docker run -it --rm -v carma-config-data:/opt/carma/vehicle/config $carma_base bash
+}
+
+
+swarm-config__status() {
+    if [ ! -z "$(docker volume ls | grep carma-config-data)" ]; then
+        echo "Found carma-config-data volume"
+    else
+        echo "Missing carma-config-data volume, add with \"carma swarm update-config <image tag>\""
+        exit 1
+    fi
+
+    if [ -z "$1" ]; then
+        echo ""
+        docker run --rm -v carma-config-data:/opt/carma/vehicle/config --entrypoint sh busybox:latest -c \
+        'cat /opt/carma/vehicle/config/docker-compose-manager.yml <(echo)
+         cat /opt/carma/vehicle/config/docker-compose-worker.yml <(echo)
+         cat /opt/carma/vehicle/config/docker-compose-background-manager.yml <(echo)
+         cat /opt/carma/vehicle/config/carma.launch <(echo) 
+         cat /opt/carma/vehicle/config/carma.config.js <(echo)
+         cat /opt/carma/vehicle/config/carma.urdf <(echo)'
+        
+    else
+        echo " -- $1:"
+        docker run --rm -v carma-config-data:/opt/carma/vehicle/config --entrypoint sh busybox:latest -c \
+        "cat /opt/carma/vehicle/config/$1"
+    fi
+
+}
+
+
+swarm__attach() {
+    echo "Attaching STDOUT logs to terminal..."
+
+    SWARM_SERVICES=$(docker service ls --format "{{.Name}}")
+
+    # this seems to loop forever and cannot be escaped with ctrl+C
+    # performance will likely be poor due to combined log streams
+    for TASK in $SWARM_SERVICES; do
+        docker service logs --follow --details --tail=10 "$TASK" &
+    done
+
+    sleep infinity
+}
+
+
+swarm__register() {
+    echo "Installing Docker images..."
+
+    if [ "$(docker volume ls | grep carma-registry | sed 's/local     //g')" ]; then
+        echo "Found carma-registry volume"
+    else
+        echo "Missing carma-registry volume, adding now..."
```

Comment:
`carma-registry` should be quoted since they are the names. It's a small visual aid that could help users.
```shell
echo "Found 'carma-registry' volume"
```